package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -47,47 +59,951 @@ ctypedef cset[int] intset
 ctypedef cmap[int, intset] intmap
 
 
-@cython.boundscheck(False) # turn off bounds-checking for entire function
-@cython.wraparound(False)  # turn off negative index wrapping for entire function
-def unique_int_1d(np.int64_t[:] values):
+cdef inline bint isinf(double x) nogil:
+    """Check if a double is ``inf`` or ``-inf``.
+
+    In contrast to ``libc.math.isfinite``, this function keeps working when
+    compiled with ``-ffast-math``.
+
+
+    .. versionadded:: 0.20.0
+    """
+    return (x == INFINITY) | (x == -INFINITY)
+
+
+def coords_add_vector(float[:, :] coordinates not None,
+                      np.ndarray vector not None):
+    """Add `vector` to each position in `coordinates`.
+
+    Equivalent to ``coordinates += vector`` but faster for C-contiguous
+    coordinate arrays. Coordinates are modified in place.
+
+    Parameters
+    ----------
+    coordinates: numpy.ndarray
+        Coordinate array of dtype ``numpy.float32`` and shape ``(n, 3)``.
+    vector: numpy.ndarray
+        Single coordinate vector of shape ``(3,)``.
+
+    Raises
+    ------
+    ValueError
+        If the shape of `coordinates` is not ``(n, 3)`` or if the shape of
+        `vector` is not ``(3,)``.
+
+
+    .. versionadded:: 0.20.0
+    """
+    cdef np.intp_t cshape1 = coordinates.shape[1]
+    cdef np.intp_t vndim = vector.ndim
+    cdef np.intp_t vshape0 = vector.shape[0]
+    if cshape1 != 3:
+        raise ValueError("Wrong shape: positions.shape != (n, 3)")
+    if vndim != 1 or vshape0 != 3:
+        raise ValueError("Wrong shape: vector.shape != (3,)")
+    if vector.dtype == np.float32:
+        _coords_add_vector32(coordinates, vector)
+    else:
+        _coords_add_vector64(coordinates, vector.astype(np.float64, copy=False))
+
+
+cdef inline void _coords_add_vector32(float[:, :]& coordinates,
+                                      float[:]& vector) nogil:
+    """Low-level implementation of :func:`coords_add_vector` for float32
+    vectors.
+
+
+    .. versionadded:: 0.20.0
+    """
+    cdef np.intp_t i
+    for i in range(coordinates.shape[0]):
+        coordinates[i, 0] += vector[0]
+        coordinates[i, 1] += vector[1]
+        coordinates[i, 2] += vector[2]
+
+
+cdef inline void _coords_add_vector64(float[:, :]& coordinates,
+                                      double[:]& vector) nogil:
+    """Low-level implementation of :func:`coords_add_vector` for non-float32
+    vectors.
+
+
+    .. versionadded:: 0.20.0
+    """
+    cdef np.intp_t i
+    for i in range(coordinates.shape[0]):
+        # Don't use += here!
+        coordinates[i, 0] = coordinates[i, 0] + vector[0]
+        coordinates[i, 1] = coordinates[i, 1] + vector[1]
+        coordinates[i, 2] = coordinates[i, 2] + vector[2]
+
+
+def _coords_add_vectors(np.ndarray[np.float32_t, ndim=2] coordinates not None,
```

Comment:
Is there a compelling reason to use the old buffer interface instead of a typed memoryview here?

Reply:
@tylerjereddy Yes, there is. Depending on input dtype and/or shape, I use numpy operations such as `ndarray.astype()` or `radds` like `coordinates += vectors`. The problem with memoryviews is that they don't offer such functionality. Having memoryviews as input would require unnecessary instantiations of `ndarray` objects with `np.asarray()` prior to these operations... You see my point?
So we really *do* want to have `numpy.ndarray` objects as input here. As you might have noticed, I also initialize a memoryview of the same data in line 203:
```python
cdef float[:, :] coords = coordinates
```
Whenever I use custom cython operations, this memoryview is used instead of the `ndarray` objects.

Reply:
I still try to guard against introduction of the old buffer interface in new code. Usually there is a way to work around using the old approach with careful design.

I agree it is a pain to deal with `asarray()` stuff a lot.

Reply:
>Usually there is a way to work around using the old approach with careful design.

Do you have anything specific in mind? Like a Python wrapper taking care of array initialization and returning? (If that's not too costly)

Reply:
Not sure, but probably not a blocking issue here if you don't have something in mind already.

I'm also not sure if Cython will ever really deprecate the old approach. I've had a hard time replacing the old buffers with memoryviews in SciPy for the reasons you outline--reviewers don't like performance drops when the buffer is just a medium for transmitting the ndarray.